synapse/util/ratelimitutils.py

```diff
@@ -176,8 +177,11 @@ def on_both(r: object) -> object:
             # Ensure that we've properly cleaned up.
             self.sleeping_requests.discard(request_id)
             self.ready_request_queue.pop(request_id, None)
+            wait_span_scope.__exit__(None, None, None)
```

Comment:
I'm slightly surprised this works, since `wait_span_scope` isn't declared at the point `on_both` is defined. But I assume you've tested it and python scope *do* in fact work this way :)

Reply:
It does work but I would hope mypy would complain if it didn't 🙏